### ISSUE TITLE
New notifications scripts will always fail in an IPv6 only environment

### ISSUE THREAD

**Issue by Bobobo-bo-Bo-bobo** [Icinga/icinga2#5470]
The new notification scripts in 2.7 (`mail-service-notification.sh` and `mail-host-notification.sh`) will always fail to notify in an IPv6 only environment.

## Expected Behavior
Mail notification will work in an IPv6 only environment.

## Current Behavior
The new notification scripts check if `HOSTADDRESS` from `$address$` is set. In an IPv6 only environment only `$address6$` (`HOSTADDRESS6`) is set. As the new scripts only check for `HOSTADDRESS` and not `HOSTADDRESS6`

```
[...]
## Check required parameters (TODO: better error message)
## Keep formatting in sync with mail-service-notification.sh
if [ ! "$HOSTADDRESS" ] || [ ! "$LONGDATETIME" ] \
|| [ ! "$HOSTNAME" ] || [ ! "$HOSTDISPLAYNAME" ] \
|| [ ! "$HOSTOUTPUT" ] || [ ! "$HOSTSTATE" ] \
|| [ ! "$USEREMAIL" ] || [ ! "$NOTIFICATIONTYPE" ]; then
  Error "Requirement parameters are missing."
fi
[...]
```

## Possible Solution
Checking for `HOSTADDRESS` and `HOSTADDRESS6`. So something like:

```
root@debian9164m /etc/icinga2 # diff -u scripts/mail-host-notification.sh.orig scripts/mail-host-notification.sh                                                                                                                                                                                                            
--- scripts/mail-host-notification.sh.orig      2017-08-07 19:25:51.718903531 +0200
+++ scripts/mail-host-notification.sh   2017-08-07 19:36:09.547170137 +0200
@@ -16,7 +16,7 @@
 cat << EOF

 Required parameters:
-  -4 HOSTADDRESS (\$address\$)
+  -4 HOSTADDRESS (\$address\$) or -6 HOSTADDRESS6 (\$address6\$)
   -d LONGDATETIME (\$icinga.long_date_time\$)
   -l HOSTNAME (\$host.name\$)
   -n HOSTDISPLAYNAME (\$host.display_name\$)
@@ -26,7 +26,6 @@
   -t NOTIFICATIONTYPE (\$notification.type\$)

 Optional parameters:
-  -6 HOSTADDRESS6 (\$address6\$)
   -b NOTIFICATIONAUTHORNAME (\$notification.author\$)
   -c NOTIFICATIONCOMMENT (\$notification.comment\$)
   -i ICINGAWEB2URL (\$notification_icingaweb2url\$, Default: unset)
@@ -81,7 +80,7 @@

 ## Check required parameters (TODO: better error message)
 ## Keep formatting in sync with mail-service-notification.sh
-if [ ! "$HOSTADDRESS" ] || [ ! "$LONGDATETIME" ] \
+if [ ! "$HOSTADDRESS" -a ! "$HOSTADDRESS6" ] || [ ! "$LONGDATETIME" ] \
 || [ ! "$HOSTNAME" ] || [ ! "$HOSTDISPLAYNAME" ] \
 || [ ! "$HOSTOUTPUT" ] || [ ! "$HOSTSTATE" ] \
 || [ ! "$USEREMAIL" ] || [ ! "$NOTIFICATIONTYPE" ]; then
```

and

```
root@debian9164m /etc/icinga2 # diff -u scripts/mail-service-notification.sh.orig scripts/mail-service-notification.sh
--- scripts/mail-service-notification.sh.orig   2017-08-07 19:35:54.611162721 +0200
+++ scripts/mail-service-notification.sh        2017-08-07 19:25:16.010890504 +0200
@@ -16,7 +16,7 @@
 cat << EOF

 Required parameters:
-  -4 HOSTADDRESS (\$address\$)
+  -4 HOSTADDRESS (\$address\$) or -6 HOSTADDRESS6 (\$address6\$)
   -d LONGDATETIME (\$icinga.long_date_time\$)
   -e SERVICENAME (\$service.name\$)
   -l HOSTNAME (\$host.name\$)
@@ -28,7 +28,6 @@
   -u SERVICEDISPLAYNAME (\$service.display_name\$)

 Optional parameters:
-  -6 HOSTADDRESS6 (\$address6\$)
   -b NOTIFICATIONAUTHORNAME (\$notification.author\$)
   -c NOTIFICATIONCOMMENT (\$notification.comment\$)
   -i ICINGAWEB2URL (\$notification_icingaweb2url\$, Default: unset)
@@ -85,12 +84,13 @@

 ## Check required parameters (TODO: better error message)
 ## Keep formatting in sync with mail-host-notification.sh
-if [ ! "$HOSTADDRESS" ] || [ ! "$LONGDATETIME" ] \
+if [ ! "$HOSTADDRESS" -a ! "$HOSTADDRESS6" ] || [ ! "$LONGDATETIME" ] \
 || [ ! "$HOSTNAME" ] || [ ! "$HOSTDISPLAYNAME" ] \
 || [ ! "$SERVICENAME" ] || [ ! "$SERVICEDISPLAYNAME" ] \
 || [ ! "$SERVICEOUTPUT" ] || [ ! "$SERVICESTATE" ] \
 || [ ! "$USEREMAIL" ] || [ ! "$NOTIFICATIONTYPE" ]; then
-  Error "Requirement parameters are missing."
+       Error "Requirement parameters are missing."
+
 fi

 ## Build the message's subject
```

## Steps to Reproduce (for bugs)
1. Disable IPv4, use IPv6 only
2. trigger the notification

## Context
Running Icinga2 in an IPv6 only environment without any IPv4 addresses.

## Your Environment
<!--- Include as many relevant details about the environment you experienced the problem in -->
* Version used (`icinga2 --version`):
```
root@debian9164m ~ # icinga2 --version
icinga2 - The Icinga 2 network monitoring daemon (version: r2.7.0-1)

Copyright (c) 2012-2017 Icinga Development Team (https://www.icinga.com/)
License GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl2.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Application information:
  Installation root: /usr
  Sysconf directory: /etc
  Run directory: /run
  Local state directory: /var
  Package data directory: /usr/share/icinga2
  State path: /var/lib/icinga2/icinga2.state
  Modified attributes path: /var/lib/icinga2/modified-attributes.conf
  Objects path: /var/cache/icinga2/icinga2.debug
  Vars path: /var/cache/icinga2/icinga2.vars
  PID path: /run/icinga2/icinga2.pid

System information:
  Platform: Debian GNU/Linux
  Platform version: 9 (stretch)
  Kernel: Linux
  Kernel version: 4.9.0-3-amd64
  Architecture: x86_64

Build information:
  Compiler: GNU 6.3.0
  Build host: debian9
```

* Operating System and version: Debian 9, x86_64

* Enabled features (`icinga2 feature list`):
```
root@debian9164m ~ # icinga2 feature list
Disabled features: compatlog debuglog gelf graphite opentsdb perfdata statusdata
Enabled features: api checker command ido-pgsql influxdb livestatus mainlog notification syslog
```

* Config validation (`icinga2 daemon -C`):
```
root@debian9164m ~ # icinga2 daemon -C
information/cli: Icinga application loader (version: r2.7.0-1)
information/cli: Loading configuration file(s).
information/ConfigItem: Committing config item(s).
information/ApiListener: My API identity: debian9164m.ypbind.de
warning/ApplyRule: Apply rule 'smtp' (in /etc/icinga2/zones.d/global-configuration/services/local_smtp.conf: 1:0-1:19) for type 'Service' does not match anywhere!
warning/ApplyRule: Apply rule 'remote_mdraid_mismatch' (in /etc/icinga2/zones.d/global-configuration/services/mdraid_mismatch.conf: 1:0-1:37) for type 'Service' does not match anywhere!
warning/ApplyRule: Apply rule 'check_nrpe' (in /etc/icinga2/zones.d/global-configuration/services/nrpe.conf: 1:0-1:67) for type 'Service' does not match anywhere!
warning/ApplyRule: Apply rule 'opentracker' (in /etc/icinga2/zones.d/global-configuration/services/opentracker.conf: 1:0-1:26) for type 'Service' does not match anywhere!
warning/ApplyRule: Apply rule 'saltminion' (in /etc/icinga2/zones.d/global-configuration/services/salt_minion.conf: 1:0-1:25) for type 'Service' does not match anywhere!
warning/ApplyRule: Apply rule 'xmpp' (in /etc/icinga2/zones.d/global-configuration/services/xmpp.conf: 1:0-1:19) for type 'Service' does not match anywhere!
information/ConfigItem: Instantiated 4 ApiUsers.
information/ConfigItem: Instantiated 1 ApiListener.
information/ConfigItem: Instantiated 1 SyslogLogger.
information/ConfigItem: Instantiated 11 Zones.
information/ConfigItem: Instantiated 1 FileLogger.
information/ConfigItem: Instantiated 10 Endpoints.
information/ConfigItem: Instantiated 4 NotificationCommands.
information/ConfigItem: Instantiated 277 Notifications.
information/ConfigItem: Instantiated 228 CheckCommands.
information/ConfigItem: Instantiated 14 Hosts.
information/ConfigItem: Instantiated 1 IcingaApplication.
information/ConfigItem: Instantiated 2 HostGroups.
information/ConfigItem: Instantiated 1 UserGroup.
information/ConfigItem: Instantiated 3 TimePeriods.
information/ConfigItem: Instantiated 1 User.
information/ConfigItem: Instantiated 263 Services.
information/ConfigItem: Instantiated 1 CheckerComponent.
information/ConfigItem: Instantiated 1 IdoPgsqlConnection.
information/ConfigItem: Instantiated 1 InfluxdbWriter.
information/ConfigItem: Instantiated 1 LivestatusListener.
information/ConfigItem: Instantiated 1 NotificationComponent.
information/ConfigItem: Instantiated 1 ExternalCommandListener.
information/ScriptGlobal: Dumping variables to file '/var/cache/icinga2/icinga2.vars'
information/cli: Finished validating the configuration file(s).
```

**Comment by dnsmichi** [comment]
I'm not sure how many users are affected by this (the majority seems to have the `address` field set ). Notification scripts are located in /etc/icinga2 and your changes won't be overridden on upgrade, so you are safe for now.

Does it make sense to entirely render IPv4 addresses optional?
https://github.com/Icinga/icinga2/blob/master/etc/icinga2/scripts/mail-service-notification.sh#L110

@Crunsher @gunnarbeutner if we change that somehow, only for v2.8.0 imho.

**Comment by Bobobo-bo-Bo-bobo** [comment]
You are right, the majority will still have IPv4 addresses because for most networks they are well suited. So this is not the "usual use case" (the fact that Icinga2 works pretty well in an IPv6 only environment was one of the reasons we switched from check_mk to Icinga2 ;-) ).

As for your question. Personally I can live with an empty value for the IPv4 address; no IPv4 address, so no value. Especially as https://github.com/Icinga/icinga2/blob/master/etc/icinga2/scripts/mail-service-notification.sh#L115 will display the IPv6 for the host/service notification message. As long as it will not refuse to run if there is no `HOSTADDRESS`.

**Comment by dnsmichi** [comment]
Merged the PR which will be included in 2.7.1, thanks.
